File: 백준/Gold/9328. 열쇠/열쇠.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    static final char EMPTY = '.';
    static final char WALL = '*';
    static final char DOCUMENT = '$';

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    static int h, w;

    static char[][] map;
    static boolean[] isGetKey;
    static ArrayList<Point>[] lazy;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine());

        StringBuilder sb = new StringBuilder();

        while (t-- > 0) {

            StringTokenizer st = new StringTokenizer(br.readLine());

            h = Integer.parseInt(st.nextToken()) + 2;
            w = Integer.parseInt(st.nextToken()) + 2;

            map = new char[h][w];
            isGetKey = new boolean[26];
            lazy = new ArrayList[26];

            for (int i = 0; i < lazy.length; i++) {
                lazy[i] = new ArrayList<>();
            }

            for (char[] row : map) {
                Arrays.fill(row, EMPTY);
            }

            for (int i = 1; i <= h - 2; i++) {
                char[] arr = br.readLine().toCharArray();
                for (int j = 1; j <= w - 2; j++) {
                    map[i][j] = arr[j - 1];
                }
            }

            char[] keys = br.readLine().toCharArray();

            int index = 0;
            while (index < keys.length && keys[index] != '0') {
                isGetKey[keys[index++] - 'a'] = true;
            }

            sb.append(bfs()).append("\n");
        }

        System.out.print(sb);
    }

    private static int bfs() {

        Queue<Point> qu = new ArrayDeque<>();
        qu.offer(new Point(0, 0));

        boolean[][] visit = new boolean[h][w];
        visit[0][0] = true;

        int count = 0;

        while (!qu.isEmpty()) {

            Point now = qu.poll();
            int x = now.x;
            int y = now.y;

            for (int i = 0; i < 4; i++) {

                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx < 0 || ny < 0 || nx >= h || ny >= w || visit[nx][ny] || map[nx][ny] == WALL) {
                    continue;
                }

                char ch = map[nx][ny];

                if (ch == DOCUMENT) {
                    count++;
                    map[nx][ny] = EMPTY;
                    visit[nx][ny] = true;
                    qu.offer(new Point(nx, ny));
                } else if (Character.isUpperCase(ch)) {
                    if (isGetKey[ch - 'A']) {
                        map[nx][ny] = EMPTY;
                        visit[nx][ny] = true;
                        qu.offer(new Point(nx, ny));
                    } else {
                        lazy[ch - 'A'].add(new Point(nx, ny));
                    }
                } else if (Character.isLowerCase(ch)) {
                    map[nx][ny] = EMPTY;
                    isGetKey[ch - 'a'] = true;

                    for (Point p : lazy[ch - 'a']) {
                        if (!visit[p.x][p.y]) {
                            visit[p.x][p.y] = true;
                            qu.offer(new Point(p.x, p.y));
                        }
                    }

                    lazy[ch - 'a'].clear();
                    visit[nx][ny] = true;
                    qu.offer(new Point(nx, ny));
                } else {
                    visit[nx][ny] = true;
                    qu.offer(new Point(nx, ny));
                }
            }
        }

        return count;
    }

    static class Point {
        int x, y;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: 백준/Gold/9328. 열쇠/열쇠.py
from collections import deque

EMPTY = '.'
WALL = '*'
DOCUMENT = '$'

dx = (-1, 1, 0, 0)
dy = (0, 0, -1, 1)
h, w = 0, 0
my_map = [[]]
isGetKey = []
lazy = []


def bfs():
    qu = deque([(0, 0)])
    visit = [[False] * w for _ in range(h)]
    visit[0][0] = True

    count = 0

    while qu:
        x, y = qu.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx < 0 or ny < 0 or nx >= h or ny >= w or visit[nx][ny] or my_map[nx][ny] == WALL:
                continue

            ch = my_map[nx][ny]

            if ch == DOCUMENT:
                count += 1
                my_map[nx][ny] = EMPTY

            if ch.isupper():
                if isGetKey[ord(ch) - ord('A')]:
                    my_map[nx][ny] = EMPTY
                else:
                    lazy[ord(ch) - ord('A')].append((nx, ny))
                    continue

            if ch.islower():
                my_map[nx][ny] = EMPTY
                isGetKey[ord(ch) - ord('a')] = True

                for px, py in lazy[ord(ch) - ord('a')]:
                    visit[px][py] = True
                    qu.append((px, py))

            visit[nx][ny] = True
            qu.append((nx, ny))

    return count


t = int(input())
ans = []
for _ in range(t):

    h, w = map(int, input().split())
    h += 2
    w += 2

    my_map = [[EMPTY] * w for _ in range(h)]
    isGetKey = [False] * 26
    lazy = [[] for _ in range(26)]

    for i in range(1, h - 1):
        row = input().strip()
        for j in range(1, w - 1):
            my_map[i][j] = row[j - 1]

    keys = input().strip()

    for key in keys:
       if key == '0':
           break
       isGetKey[ord(key) - ord('a')] = True

    ans.append(str(bfs()))

print("\n".join(ans))
